src/Microsoft.DotNet.ProjectJsonMigration/transforms/TransformApplicator.cs

```diff
@@ -0,0 +1,170 @@
+                    var mergeResult = MergeItems(item, existingItem);
+                    item = mergeResult.InputItem;
+
+                    // Existing Item is null when it's entire set of includes has been merged with the MergeItem
+                    if (mergeResult.ExistingItem == null)
+                    {
+                        existingItem.Parent.RemoveChild(existingItem);
+                    }
+
+                    Execute(mergeResult.MergedItem, destinationItemGroup);
+                }
+
+                // Item will be null only when it's entire set of includes is merged with existing items
+                if (item != null)
+                {
+                    Execute(item, destinationItemGroup);
+                }
+            }
+            else
+            {
+                Execute(item, destinationItemGroup);
+            }
+        }
+
+        public void Execute(
+            IEnumerable<ProjectItemElement> items,
+            ProjectItemGroupElement destinationItemGroup,
+            bool mergeExisting)
+        {
+            foreach (var item in items)
+            {
+                Execute(item, destinationItemGroup, mergeExisting);
+            }
+        }
+
+        /// <summary>
+        /// Merges two items on their common sets of includes.
+        /// The output is 3 items, the 2 input items and the merged items. If the common
+        /// set of includes spans the entirety of the includes of either of the 2 input
+        /// items, that item will be returned as null.
+        ///
+        /// The 3rd output item, the merged item, will have the Union of the excludes and
+        /// metadata from the 2 input items. If any metadata between the 2 input items is different,
+        /// this will throw.
+        ///
+        /// This function will mutate the Include property of the 2 input items, removing the common subset.
+        /// </summary>
+        private MergeResult MergeItems(ProjectItemElement item, ProjectItemElement existingItem)
+        {
+            if (!string.Equals(item.ItemType, existingItem.ItemType, StringComparison.Ordinal))
+            {
+                throw new InvalidOperationException("Cannot merge items of different types.");
+            }
+
+            if (!item.CommonIncludes(existingItem).Any())
+            {
+                throw new InvalidOperationException("Cannot merge items without a common include.");
+            }
+
+            var commonIncludes = item.CommonIncludes(existingItem).ToList();
+            item.RemoveIncludes(commonIncludes);
+            existingItem.RemoveIncludes(commonIncludes);
+
+            var mergedItem = _projectElementGenerator.AddItem(item.ItemType, string.Join(";", commonIncludes));
+
+            mergedItem.AddExcludes(existingItem.Excludes());
+            mergedItem.AddExcludes(item.Excludes());
+
+            mergedItem.AddMetadata(existingItem.Metadata);
+            mergedItem.AddMetadata(item.Metadata);
```

Comment:
I may not have a full understanding of this, but in this situation:
```
<ItemGroup>
  <X Include="x">
    <M>y</M>
  </X>
  <X Include="x">
    <N>z</N>
  </X>
</ItemGroup>
```

my expectation is that because both `X` items have the same include, they have the same identity and there would be effectively a single `X` item with two metadata items. Is there a gap in my understanding there?

Avoiding this situation where we have duplicate includes between two items is the intention of this code, as when that was done in items like `Compile` things would blow up.